### PR TITLE
Make Keycloak FIPS working with OpenJDK 17 on FIPS enabled RHEL

### LINKED ISSUE #15721: Doublecheck BCFIPS setup on OpenJDK 17
Especially the setup with just BCFIPS providers (Setup without `SUN` provider). Here are some details why BCFIPS fail at startup: https://github.com/bcgit/bc-java/issues/1285

The root cause is the fact that `SecureRandom.getInstanceStrong()` is not available in OpenJDK 17 on the FIPS enabled RHEL. Details in: https://bugzilla.redhat.com/show_bug.cgi?id=2155060

### PR DESCRIPTION
**Pull request by mposolda**
Closes #15721

### CODE REVIEW

**Comment by mposolda** [comment]
@pedroigor @stianst Anyone available for the review of the OpenJDK 17 support for FIPS?

Few things done in this PR:

- There is a workaround for the OpenJDK 17 bug https://bugzilla.redhat.com/show_bug.cgi?id=2155060 (or at least I think it is a bug, not yet confirmed from the OpenJDK team). BCFIPS doesn't work on OpenJDK 17 due to this.

- Excluded `org.bouncycastle:bcutil-jdk15on` from the classpath (This was accidentally still present on classpath for the FIPS mode, which causes issues on OpenJDK 17)

- Moreover it is needed that default security providers from the `java.security` file are available on OpenJDK 17. The setup with only BCFIPS security providers available during bootstrap  (Step 4 from the file https://github.com/mposolda/keycloak/blob/15721-fips-openjdk17/docs/fips.md#run-the-server-with-fips) doesn't work on OpenJDK 17. On the other hand, this is fine as OpenJDK 17 on the RHEL FIPS explicitly disables usage of the non-FIPS compliant algorithms from SUN provider and hence there is no risk of accidentally using those in the application (This was not the case with OpenJDK 11 where the non-secured algorithms were allowed to be used). I am considering removing Step 4 entirely from fips.md if we stick only to OpenJDK 17 support (no OpenJDK 11) once we remove support for OpenJDK 11.

- Few more tweaks in the tests (For example to have KcReg and KcAdm tests working with OpenJDK 17 with the removed bcutil-jdk15on from classpath)

File: crypto/fips1402/src/main/java/org/keycloak/crypto/fips/FIPS1402Provider.java [hunk 1]
```diff
@@ -262,4 +259,34 @@ private SNIHostName getSNIHostName(String host) {
...
+    private void checkSecureRandom(Runnable insertBcFipsProvider) {
+        try {
+            SecureRandom sr = SecureRandom.getInstanceStrong();
+            log.debugf("Strong secure random available. Algorithm: %s, Provider: %s", sr.getAlgorithm(), sr.getProvider());
+            insertBcFipsProvider.run();
+        } catch (NoSuchAlgorithmException nsae) {
+
+            // Fallback to regular SecureRandom
+            SecureRandom secRandom = new SecureRandom();
+            String origStrongAlgs = Security.getProperty("securerandom.strongAlgorithms");
```
Comment (pedroigor) [inline crypto/fips1402/src/main/java/org/keycloak/crypto/fips/FIPS1402Provider.java:273]: Is this property not set by any chance? If you look at the `SecureRandom` impl it should always fail with `NoSuchAlgorithmException`  when calling `getInstanceStrong` if the property is not set.

File: crypto/fips1402/src/main/java/org/keycloak/crypto/fips/FIPS1402Provider.java [hunk 2]
```diff
@@ -262,4 +259,34 @@ private SNIHostName getSNIHostName(String host) {
 
         };
     }
+
+    // BCFIPS require "SecureRandom.getInstanceStrong" to be available. But it may not be available on RHEL 8 on OpenJDK 17 due the https://bugzilla.redhat.com/show_bug.cgi?id=2155060
```
Comment (pedroigor) [inline crypto/fips1402/src/main/java/org/keycloak/crypto/fips/FIPS1402Provider.java:263]: I'm wondering if this is actually a bug or behavior we need to take into account when running on RHEL. See https://access.redhat.com/documentation/en-us/openjdk/17/pdf/configuring_openjdk_17_on_rhel_with_fips/openjdk-17-configuring_openjdk_17_on_rhel_with_fips-en-us.pdf.

I'm not sure about this, but looks like you can use `java.security.disableSystemPropertiesFile` to disable global FIPS and then force a specific algorithm.

File: crypto/fips1402/src/main/java/org/keycloak/crypto/fips/FIPS1402Provider.java (hunk 1 above)
Comment (mposolda) [reply crypto/fips1402/src/main/java/org/keycloak/crypto/fips/FIPS1402Provider.java:273]: There is also another possibility how the `NoSuchAlgorithmException` can be thrown from the `getInstanceStrong` - when property `"securerandom.strongAlgorithms` is specified, but it's pointing to the algorithms, which are not available or disabled. That is the case here.

This property is set by default in `java.security` on all the environments I saw so far. And it is typically set in OpenJDK by default to `NativePRNGBlocking:SUN,DRBG:SUN` (applies for both my Ubuntu as well as for RHEL with or without FIPS enabled, and for both OpenJDK11 and OpenJDK17). The problem is that none of the specified algorithms `NativePRNGBlocking:SUN,DRBG:SUN` is available on the OpenJDK 17 on the FIPS enabled RHEL. That's also clear fro m the exception message (For example see here how the exception looks like: https://bugzilla.redhat.com/show_bug.cgi?id=2155060)

I think that it is a bug as `getInstanceStrong` should be available per javadoc of this method. The `SecureRandom.getInstanceStrong()` is used by BCFIPS and BCFIPS fails to bootstrap when this method is not available. Also I talked about this with the guy from BouncyCastle team and he confirmed that `getInstanceStrong` unavailable looks like a bug (and suggested to use exactly the workaround like this).

You can check https://bugzilla.redhat.com/show_bug.cgi?id=2155060 for the details (I don't yet have response from the OpenJDK team here).

File: crypto/fips1402/src/main/java/org/keycloak/crypto/fips/FIPS1402Provider.java (hunk 2 above)
Comment (mposolda) [reply crypto/fips1402/src/main/java/org/keycloak/crypto/fips/FIPS1402Provider.java:263]: The `java.security.disableSystemPropertiesFile` is applied at bootstrap of JVM and it completely disables FIPS for the JDK (Similar to `com.redhat.fips`). Using this property is not an option as we won't be FIPS compliant when it is used.

**Review by pedroigor — APPROVED** [review]

**Comment by mposolda** [comment]
@pedroigor Thanks Pedro!

**Pull request merged**